src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java

```diff
@@ -0,0 +1,262 @@
+ * resources derived from that scope (e.g. attempting to access a {@link MemorySegment} instance) will fail with {@link IllegalStateException}.
+ * <p>
+ * Closing a resource scope will cause all the cleanup actions associated with that scope (see {@link #addOnClose(Runnable)}) to be called.
+ * Moreover, closing a resource scope might trigger the releasing of the underlying memory resources associated with said scope; for instance:
+ * <ul>
+ *     <li>closing the scope associated with a native memory segment results in <em>freeing</em> the native memory associated with it
+ *     (see {@link MemorySegment#allocateNative(long, ResourceScope)}, or {@link SegmentAllocator#arenaUnbounded(ResourceScope)})</li>
+ *     <li>closing the scope associated with a mapped memory segment results in the backing memory-mapped file to be unmapped
+ *     (see {@link MemorySegment#mapFile(Path, long, long, FileChannel.MapMode, ResourceScope)})</li>
+ *     <li>closing the scope associated with an upcall stub results in releasing the stub
+ *     (see {@link CLinker#upcallStub(MethodHandle, FunctionDescriptor, ResourceScope)}</li>
+ * </ul>
+ *
+ * <h2>Implicit deallocation</h2>
+ *
+ * Resource scopes can be associated with a {@link Cleaner} instance (see {@link #ofConfined(Cleaner)}) - we call these
+ * resource scopes <em>managed</em> resource scopes. A managed resource scope is closed automatically once the scope instance
+ * becomes <a href="../../../java/lang/ref/package.html#reachability">unreachable</a>.
+ * <p>
+ * Managed resource scopes can still be closed explicitly (see {@link #close()}); this can be useful to allow for predictable,
+ * deterministic resource deallocation, while still prevent accidental native memory leaks. In case a managed resource
+ * scope is closed explicitly, no further action will be taken when the scope becomes unreachable; that is, cleanup actions
+ * (see {@link #addOnClose(Runnable)}) associated with a resource scope, whether managed or not, are called <em>exactly once</em>.
+ *
+ * <h2><a id = "thread-confinement">Thread confinement</a></h2>
+ *
+ * Resource scopes can be further divided into two categories: <em>thread-confined</em> resource scopes, and <em>shared</em>
+ * resource scopes.
+ * <p>
+ * Confined resource scopes (see {@link #ofConfined()}), support strong thread-confinement guarantees. Upon creation,
+ * they are assigned an <em>owner thread</em>, typically the thread which initiated the creation operation (see {@link #ownerThread()}).
+ * After creating a confined resource scope, only the owner thread will be allowed to directly manipulate the resources
+ * associated with this resource scope. Any attempt to perform resource access from a thread other than the
+ * owner thread will result in a runtime failure.
+ * <p>
+ * Shared resource scopes (see {@link #ofShared()}), support strong thread-confinement guarantees. A shared resource scope
```

Comment:
Noticed from the generated javadoc; is this a typo? "support string thread-confinement guarantees".